sound/soc/sof/intel/hda-dai.c

```diff
@@ -56,13 +63,22 @@ hda_dai_get_ops(struct snd_pcm_substream *substream, struct snd_soc_dai *cpu_dai
 	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(cpu_dai, substream->stream);
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(cpu_dai->component);
 	struct snd_sof_widget *swidget = w->dobj.private;
-	struct snd_sof_dai *sdai = swidget->private;
+	const struct hda_dai_widget_dma_ops *ops;
```

Comment:
@ujfalusi I think something along these lines as below might look better. What do you think?
```
static const struct hda_dai_widget_dma_ops *
hda_dai_get_ops(struct snd_pcm_substream *substream, struct snd_soc_dai *cpu_dai)
{
	struct snd_soc_dapm_widget *w = snd_soc_dai_get_widget(cpu_dai, substream->stream);
	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(cpu_dai->component);
	struct snd_sof_widget *swidget = w->dobj.private;
	const struct hda_dai_widget_dma_ops *ops;
	struct snd_sof_dai *sdai;

	/* return already set ops */
	if (swidget) {
		sdai = swidget->private;
		if (sdai->platform_private)
			return sdai->platform_private;
	}

	/* select the op */
	ops = hda_select_dai_widget_ops(sdev, swidget);
	if (!ops)
		return NULL;

	/* check if mandatory ops are set */
	if (!ops || !ops->get_hext_stream)
		return NULL;

	/* set the ops for DSP mode */
	if (swidget)
		sdai->platform_private = ops;
	
	return ops;
}
```

Reply:
I have thought about the second way to allocate the swidget, let me test on Monday if it brings other issues or not.

Reply:
Right, so this is not going to work without doing a full blown topology parsing in case of DSPless mode as we need to set up all the things to get the information (`ipc4_copier->dai_type` or `dai_config->type`)
I think I will replace the `!swidget` checks with explicit `dspless_mode_selected` queries to not obscure the code.
Yep, less generic, but easier to understand.